### ISSUE TITLE
Bluesky deployment 2022-04-15

### ISSUE THREAD

**Issue by JunAishima** [NSLS-II-FMX/profile_collection#16]
with @AbbyGi 

Workstations for bsui - ws1, ws4 - ws1 never seems to be up though
IOCs: srv2, ioc1, ioc2

host: xf17id2-srv2 - /epics/iocs/notebook
Location of conda env(s) for notebooks: /epics/iocs/notebook/conda_envs
location(s) of profile collection for notebooks: /epics/iocs/notebook/.ipython/profile_collection/startup
No workstations are normally used (exclusively through notebooks in normal operation), but bsui should be tested - ws1 or ws4 are the good ones to test bsui on
Key issue with notebooks and live plots - do not use JupyterLab
Dummy motors will be used for testing - ideal for a shutdown like now

Notebook: 2022_commissioning/2022-04-15 Bluesky deployment to be used for acceptance testing

 * Run through the entire notebook - text should be printed, energy should be printed - starting environment is "Data Collection (2022-1.0)". host is xf17id2-srv2:17000
 * Rocking curve should show live plots - adding imports so that this happens (multiple "matplotlib notebook" startups) 
 * all of the code above importing dummy motors is what we must run
   * these include rocking curve and setE_motors_FMX - but these will probably be required after beam is back (hence out of scope for this work)
 * the code below importing dummy motors is for debugging purposes (of plots especially) in case the stuff above doesn't work

**Comment by AbbyGi** [comment]
Stopping at profile_collection maintenance point 7. Profile collection has been cloned into the shared location but may need to be updated (pulled) if the `/epics/iocs/notebook/.ipython/profile_collection` gets any updates before we can continue with the links.

**Comment by JunAishima** [comment]
continuing on with "Update Bluesky config..."
* bsui not available from ws1, ws4. need config for srv2 as well - Dan ran conda-base, no problems.
* checked by sourcing /etc/profile.d/bluesky.sh and results are as expected!
* oops - some of the environment variables were pointing to strange ones - BS_PYTHONPATH from Josh's beamline and BS_ENV and BS_AN_ENV to 2022-2.0.
* fixed! all ok!

**Comment by JunAishima** [comment]
Notebook stuff
* xf17id2-srv2 - /usr/share/jupyter/kernels
* change username - /etc/systemd/system/jupyter-fmx.service User and Group from softioc to softioc-fmx
* Do similar steps to what is done for manage-iocs services
* backup iocs
  * the service is called notebook - /epics/iocs/notebook
  * backup all /epics/iocs into /epics (12G)
  *  change ownership of /epics, /epics/iocs, and all enclosed services from manage-iocs to softioc-tla:n2sn-instadmin

Where are all of the services running? we will need to run softioc to softioc-tla on all of these later
* srv2
* ioc1
* ioc2

Continued
* Run softioc to softioc-tla role on srv2
* verified that it worked for /epics, /epics/iocs, all individual services
* dzdo chown softioc-fmx:n2sn-instadmin notebook
* also do all of the other files and directories
* add g+w permission - chmod g+x for notebook directory
* chmod g+s for notebook directory
* .ipython looks pretty clean
* remove a bunch of config and other files
  * .ipython
  * .jupyter

Fix permissions of directories in /epics/iocs/notebook/.ipython/profile_collection that are only w for g.
find - change directories to all have +x, files - change files to be 664

/epics/iocs/notebook/notebooks/services - jupyter-fmx.service - change to softioc-fmx:n2sn-inststaff. git committed locally, not pushed
* /usr/share/jupyter/kernels-available - create a new directory, copy a kernel.json file and modify it as necessary
* kernel code located at github.com/NSLS-II/kernels
* make symlink from /usr/share/jupyter to //epics/iocs/notebook/notebooks/kernels - the latter is version controlled
* 

Future work:
* /epics/iocs/notebook - change git remote to a different place - gitlab

**Comment by JunAishima** [comment]
notebook not starting
* go to /var/log/jupyter and modify permissions as necessary
* kernel 2022-2.0 available - select. seems to start up
* we can get help, get_energy()
* modify profile_collection and see whether restarting will show the change
* behavior change - bec has disabled plotting by default
  * bec.enable_plots()
* make symlink from /nsls2/data profile collection to /epics/iocs/notebook/.ipython/profile_collection - see if it works!
  * enable service
* kafka test - it works! from ws1, ipython?
* bsui - starts up after using the symlinked profile collection - latest tiled environment
* plotting funny on IOS - trying to find out what the problem is - pylab 'auto' - disable

**Comment by JunAishima** [comment]
note to self: rescue PersistentDict - from GPFS
* done - moved to /nsls2/data/fmx/shared/config
  * permission changed to 660
* PersistentDict location in 01-bluesky.py changed to above
* bsui now starts up!
* history.sqlite is *not* in version control, though it is in the profile_collection tree

**Comment by JunAishima** [comment]
* /home/xf17id2 directory made
* server password setup - see n2sn-inststaff-fmx "Jupyter server"
* run role to update kafka password - "Bluesky - Deploy Kafka Config" - xf17id2-srv2 and xf17id1-ca1
* persistent dict permissions fixed (via copying to a fresh directory)
* kafka publishing tested - working!
* databroker startup issues - 2022-2.2-py39 has databroker 2.0.0b, which is considered less than 2. don't want that!
* note that jupyter logging was NOT helpful in location the persistentDict permission issue
* ownership of /usr/share/jupyter changed to xf17id2
* added password for the service - in Bitwarden under n2sn-inststaff-fmx

acceptance test working, plotting showing in scan

**Comment by JunAishima** [comment]
ws1, ws4 (have to log in via ssh.nsls2.bnl.gov), srv2 all have been updated to 2.2. bsui starts up on ws1. ioc1, ioc2, srv2 all have softioc-fmx services running.

Looks ok. Done!
